### PR TITLE
sfs/writer: Error handling and recovery

### PR DESCRIPTION
**Pull request by irq0**
Depends on  #124
Depends on  #135 

Replace iostream code with cstyle alternatives that don't throw and are closer to how ceph src/os does IO. Handle errors during open/close/write by deleting the failed file and returning an error.

Convert errno errors to S3 errors where it makes sense.

fsync directories after creating deleting the object file.

https://github.com/aquarist-labs/s3gw/issues/363

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

### CODE REVIEW

**Comment by irq0** [comment]
Looking for some general feedback and there is one thing open:
What to do with the database metadata on failure?
Right now the patch deletes the data, but keeps the metadata in state WRITING

1. keep it - let the GC remove partial objects. mark "FAILED"
2. remove on failure
3. only create after the write is finished

I tend to favor 1., because we could easily display a list of failed uploads and their reasons on a status page

File: src/rgw/driver/sfs/writer.cc
Comment (jecluis) [inline src/rgw/driver/sfs/writer.cc:105]: maybe assert before that `fd >= 0`?

File: src/rgw/driver/sfs/writer.cc
Comment (jecluis) [inline src/rgw/driver/sfs/writer.cc]: if we're logging failures, might be better to log them to level `-1`. Would be nice to have defined an `lsfs_derr()` macro, but given that's not available, logging to `-1` would suffice. Otherwise these errors will be hidden most of the time in a production environment.

File: src/rgw/driver/sfs/writer.cc
Comment (jecluis) [inline src/rgw/driver/sfs/writer.cc]: why are we using `fmt::format()` pretty much everywhere else, but falling back to the `<<` operator here?

**Review by jecluis — CHANGES_REQUESTED** [review]
Generally looks good, but I think the errors need to be logged at `-1` instead of `10` so they are not lost when running in production environments.

Also needs tests.

**Comment by jecluis** [comment]
> Looking for some general feedback and there is one thing open: What to do with the database metadata on failure? Right now the patch deletes the data, but keeps the metadata in state WRITING
> 
> 1. keep it - let the GC remove partial objects. mark "FAILED"
> 2. remove on failure
> 3. only create after the write is finished
> 
> I tend to favor 1., because we could easily display a list of failed uploads and their reasons on a status page

I would agree, although in case of an `ENOSPC` that might not be feasible. I'm not sure how we're going to update the DB if an `ENOSPC` is hit. Hopefully we stop writing once the disk is at `ENOSPC - some%` to avoid a situation where we're no longer able to write/rm stuff, but from the looks of this patch set it seems to me that if we hit `ENOSPC`, then it's because the underlying system call is throwing `ENOSPC` for reals.

**Comment by irq0** [comment]
I'm putting this on hold until #124 is in

**Comment by jecluis** [comment]
> I'm putting this on hold until #124 is in

making it a draft until then.

File: src/rgw/driver/sfs/writer.cc
Comment (irq0) [reply src/rgw/driver/sfs/writer.cc:105]: :+1:

File: src/rgw/driver/sfs/writer.cc
Comment (irq0) [reply src/rgw/driver/sfs/writer.cc]: :+1:

File: src/rgw/driver/sfs/writer.cc
Comment (irq0) [reply src/rgw/driver/sfs/writer.cc]: no reason. Adding a commit that converts the longer log lines to fmt

**Comment by irq0** [comment]
> Also needs tests.

Unit testing would be painful, as we would need to mock / fake the io functions, no?

We probably get most bang for buck with adding a test suite that uses an error injection filesystem. We need to do this eventually to at minimum ensure that we don't crash under IO errors, etc.

Charybdefs has a nice example that fits the ENOSPC our use case that illustrates what I mean: https://github.com/scylladb/charybdefs/blob/master/tests/out_of_disk_space_test.py

**Comment by jecluis** [comment]
I'll be happy to re-review this once the other 2 PRs are merged.

**Comment by jhmarina** [comment]
All the PRs mentioned have been merged. @jecluis @irq0 is this ready for a re-review?

**Comment by irq0** [comment]
> All the PRs mentioned have been merged. @jecluis @irq0 is this ready for a re-review?

Yes

**Review by jecluis — APPROVED** [review]
lgtm but with a caveat: IMO we should cleanup the written file only when we are the ones creating it. If the file already existed (and assuming no versioning enabled), we should not be removing the file during cleanup.

**Pull request merged**
